File: 112. Path Sum.scala
/*
Given a binary tree and a sum, determine if the tree has a root-to-leaf path such that adding up all the values along the path equals the given sum.

For example:
Given the below binary tree and sum = 22,
              5
             / \
            4   8
           /   / \
          11  13  4
         /  \      \
        7    2      1
return true, as there exist a root-to-leaf path 5->4->11->2 which sum is 22.


*/


def hasPathSum(root: TreeNode, sum: Int): Boolean = {
    if(root==null) return false;
    if(root.left==null && root.right==null && sum-root.value==0) return true; 
    if(hasPathSum(root.left,sum-root.value) || hasPathSum(root.right,sum-root.value)) return true;
    return false;
}

File: 112. Path Sum.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def hasPathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: bool
        """
        if not root:
            return False
        if root and not root.left and not root.right and sum==root.val :
            return True
        left,right = False,False
        if root.left:
            left = self.hasPathSum(root.left,sum-root.val)
        if root.right:
            right = self.hasPathSum(root.right,sum-root.val)
        return left or right
